pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,390 @@
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "clientid9",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{false, "clientid9"},
+		},
+		{
+			desc:     "Unsubscribe from a topic with an ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "clientid4",
+			err:      nil,
+			pubsub:   false,
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Unsubscribe from same topic with different ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "clientid9",
+			err:      nil,
+			pubsub:   false,
+			handler:  handler{false, "clientid9"},
+		},
+		{
+			desc:     "Unsubscribe from a non-existent topic with an ID",
+			topic:    "h",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Unsubscribe from an already unsubscribed topic with an ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{false, "clientidd4"},
+		},
+		{
+			desc:     "Unsubscribe from a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   false,
+			handler:  handler{false, "clientidd4"},
+		},
+		{
+			desc:     "Unsubscribe from an already unsubscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Unsubscribe from an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			pubsub:   false,
+			handler:  handler{false, "clientid4"},
+		},
+		{
+			desc:     "Unsubscribe from a topic with empty ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			pubsub:   false,
+			handler:  handler{false, ""},
+		},
+	}
+	for _, tc := range cases {
+		switch tc.pubsub {
+		case true:
+			err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+			assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		default:
+			err := pubsub.Unsubscribe(tc.clientID, tc.topic)
+			assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		}
+	}
+}
+
+type handler struct {
+	fail      bool
```

Comment:
This field seems to be unused since it's always false in tests.

Reply:
This is to test failing handler. Useful in TestUnsubscribe. But mqtt broker's Unsubscrribe() function behaves differently than nats and rabbitmq.
I left it here, to make things uniform across all tests, and leave room to change this functionality later.

Reply:
No need to keep it just for the sake of consistency between tests. If it's useful for the `Unsubscribe` test, why don't we use it?

Reply:
I added the cancel field to subscription. Now, like we nats and rabbitmq. We have an extra step(function call) before unsubscribing mqtt client-topic.
I've also added the testcases for this.